2.4.7 build error against GCC 15 or 16 *and* PostgreSQL 19

Hi,

2.4.7 fails to build against GCC 15 (Fedora 43) or GCC 16 (Fedora 44) and against PostgreSQL 19 (see other commentbelow BTW). Below are the logs:
 
```
gcc -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Werror=vla -Wmissing-format-attribute -Wold-style-declaration -Wimplicit-fallthrough=5 -Wcast-function-type -Wshadow=compatible-local -Wformat-security -Wmissing-variable-declarations -fno-strict-aliasing -fwrapv -fexcess-precision=standard -Wno-format-truncation -Wno-stringop-truncation -g -O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=3 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -march=x86-64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -mtls-dialect=gnu2 -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer -Wstrict-prototypes -Wold-style-definition -fPIC -fvisibility=hidden -fPIC -fvisibility=hidden -I/usr/pgsql-19/include -I/var/lib/pgsql/git/pgrpms/rpm/redhat/main/non-common/pglogical/mIn file included from pglogical_worker.h:18,
                 from pglogical_manager.c:33:
pglogical.h:143:1: warning: ‘maybe_unused’ attribute ignored [-Wattributes]
  143 | pg_attribute_printf(1, 2) pg_attribute_unused() static inline void VALGRIND_PRINTF(const char *format, ...) {}
      | ^~~~~~~~~~~~~~~~~~~
pglogical.h:143:49: error: expected identifier or ‘(’ before ‘static’
  143 | pg_attribute_printf(1, 2) pg_attribute_unused() static inline void VALGRIND_PRINTF(const char *format, ...) {}
      |                                                 ^~~~~~
ain/pglogical_19-2.4.7-build/pglogical-REL2_4_7/compat19 -Werror=implicit-function-declaration -I. -I./ -I/usr/pgsql-19/include/server -I/usr/pgsql-19/include/internal -D_GNU_SOURCE  -I/usr/include -I/usr/include/libxml2 -DWITH_GZFILEOP     -c -o compat19/pglogical_compat.o compat19/pglogical_compat.c
make[1]: *** [<builtin>: pglogical_monitoring.o] Error 1
make[1]: *** Waiting for unfinished jobs....
make[1]: *** [<builtin>: pglogical_rpc.o] Error 1
In file included from pglogical_queue.c:49:
pglogical.h:143:1: warning: ‘maybe_unused’ attribute ignored [-Wattributes]
  143 | pg_attribute_printf(1, 2) pg_attribute_unused() static inline void VALGRIND_PRINTF(const char *format, ...) {}
      | ^~~~~~~~~~~~~~~~~~~
pglogical.h:143:49: error: expected identifier or ‘(’ before ‘static’
  143 | pg_attribute_printf(1, 2) pg_attribute_unused() static inline void VALGRIND_PRINTF(const char *format, ...) {}
      |                                                 ^~~~~~
In file included from pglogical_output_config.c:23:
pglogical.h:143:1: warning: ‘maybe_unused’ attribute ignored [-Wattributes]
  143 | pg_attribute_printf(1, 2) pg_attribute_unused() static inline void VALGRIND_PRINTF(const char *format, ...) {}
      | ^~~~~~~~~~~~~~~~~~~
pglogical.h:143:49: error: expected identifier or ‘(’ before ‘static’
  143 | pg_attribute_printf(1, 2) pg_attribute_unused() static inline void VALGRIND_PRINTF(const char *format, ...) {}
      |                                                 ^~~~~~
make[1]: *** [<builtin>: pglogical_queue.o] Error 1
make[1]: *** [<builtin>: pglogical_manager.o] Error 1
make[1]: *** [<builtin>: pglogical_output_config.o] Error 1
In file included from pglogical_relcache.c:28:
pglogical.h:143:1: warning: ‘maybe_unused’ attribute ignored [-Wattributes]
  143 | pg_attribute_printf(1, 2) pg_attribute_unused() static inline void VALGRIND_PRINTF(const char *format, ...) {}
      | ^~~~~~~~~~~~~~~~~~~
pglogical.h:143:49: error: expected identifier or ‘(’ before ‘static’
  143 | pg_attribute_printf(1, 2) pg_attribute_unused() static inline void VALGRIND_PRINTF(const char *format, ...) {}
      |                                                 ^~~~~~
In file included from pglogical_repset.c:50:
pglogical.h:143:1: warning: ‘maybe_unused’ attribute ignored [-Wattributes]
  143 | pg_attribute_printf(1, 2) pg_attribute_unused() static inline void VALGRIND_PRINTF(const char *format, ...) {}
      | ^~~~~~~~~~~~~~~~~~~
pglogical.h:143:49: error: expected identifier or ‘(’ before ‘static’
  143 | pg_attribute_printf(1, 2) pg_attribute_unused() static inline void VALGRIND_PRINTF(const char *format, ...) {}
      |                                                 ^~~~~~
In file included from pglogical_worker.h:18,
                 from pglogical_apply_heap.c:71:
pglogical.h:143:1: warning: ‘maybe_unused’ attribute ignored [-Wattributes]
  143 | pg_attribute_printf(1, 2) pg_attribute_unused() static inline void VALGRIND_PRINTF(const char *format, ...) {}
      | ^~~~~~~~~~~~~~~~~~~
pglogical.h:143:49: error: expected identifier or ‘(’ before ‘static’
  143 | pg_attribute_printf(1, 2) pg_attribute_unused() static inline void VALGRIND_PRINTF(const char *format, ...) {}
      |                                                 ^~~~~~
make[1]: *** [<builtin>: pglogical_relcache.o] Error 1
make[1]: *** [<builtin>: pglogical_repset.o] Error 1
make[1]: *** [<builtin>: pglogical_apply_heap.o] Error 1
In file included from pglogical_sequences.c:35:
pglogical.h:143:1: warning: ‘maybe_unused’ attribute ignored [-Wattributes]
  143 | pg_attribute_printf(1, 2) pg_attribute_unused() static inline void VALGRIND_PRINTF(const char *format, ...) {}
      | ^~~~~~~~~~~~~~~~~~~
pglogical.h:143:49: error: expected identifier or ‘(’ before ‘static’
  143 | pg_attribute_printf(1, 2) pg_attribute_unused() static inline void VALGRIND_PRINTF(const char *format, ...) {}
      |                                                 ^~~~~~
In file included from pglogical_worker.h:18,
                 from pglogical_node.c:42:
pglogical.h:143:1: warning: ‘maybe_unused’ attribute ignored [-Wattributes]
  143 | pg_attribute_printf(1, 2) pg_attribute_unused() static inline void VALGRIND_PRINTF(const char *format, ...) {}
      | ^~~~~~~~~~~~~~~~~~~
pglogical.h:143:49: error: expected identifier or ‘(’ before ‘static’
  143 | pg_attribute_printf(1, 2) pg_attribute_unused() static inline void VALGRIND_PRINTF(const char *format, ...) {}
      |                                                 ^~~~~~
make[1]: *** [<builtin>: pglogical_sequences.o] Error 1
In file included from pglogical_worker.h:18,
                 from pglogical.c:57:
pglogical.h:143:1: warning: ‘maybe_unused’ attribute ignored [-Wattributes]
  143 | pg_attribute_printf(1, 2) pg_attribute_unused() static inline void VALGRIND_PRINTF(const char *format, ...) {}
      | ^~~~~~~~~~~~~~~~~~~
pglogical.h:143:49: error: expected identifier or ‘(’ before ‘static’
  143 | pg_attribute_printf(1, 2) pg_attribute_unused() static inline void VALGRIND_PRINTF(const char *format, ...) {}
      |                                                 ^~~~~~
pglogical.c: In function ‘pglogical_supervisor_main’:
pglogical.c:683:9: error: implicit declaration of function ‘VALGRIND_PRINTF’ [-Wimplicit-function-declaration]
  683 |         VALGRIND_PRINTF("PGLOGICAL: supervisor\n");
      |         ^~~~~~~~~~~~~~~
make[1]: *** [<builtin>: pglogical.o] Error 1
make[1]: *** [<builtin>: pglogical_node.o] Error 1
In file included from pglogical_worker.h:18,
                 from pglogical_functions.c:81:
pglogical.h:143:1: warning: ‘maybe_unused’ attribute ignored [-Wattributes]
  143 | pg_attribute_printf(1, 2) pg_attribute_unused() static inline void VALGRIND_PRINTF(const char *format, ...) {}
      | ^~~~~~~~~~~~~~~~~~~
pglogical.h:143:49: error: expected identifier or ‘(’ before ‘static’
  143 | pg_attribute_printf(1, 2) pg_attribute_unused() static inline void VALGRIND_PRINTF(const char *format, ...) {}
      |                                                 ^~~~~~
make[1]: *** [<builtin>: pglogical_functions.o] Error 1
In file included from pglogical_executor.c:61:
pglogical.h:143:1: warning: ‘maybe_unused’ attribute ignored [-Wattributes]
  143 | pg_attribute_printf(1, 2) pg_attribute_unused() static inline void VALGRIND_PRINTF(const char *format, ...) {}
      | ^~~~~~~~~~~~~~~~~~~
pglogical.h:143:49: error: expected identifier or ‘(’ before ‘static’
  143 | pg_attribute_printf(1, 2) pg_attribute_unused() static inline void VALGRIND_PRINTF(const char *format, ...) {}
      |                                                 ^~~~~~
make[1]: *** [<builtin>: pglogical_executor.o] Error 1
In file included from pglogical_dependency.c:95:
pglogical.h:143:1: warning: ‘maybe_unused’ attribute ignored [-Wattributes]
  143 | pg_attribute_printf(1, 2) pg_attribute_unused() static inline void VALGRIND_PRINTF(const char *format, ...) {}
      | ^~~~~~~~~~~~~~~~~~~
pglogical.h:143:49: error: expected identifier or ‘(’ before ‘static’
  143 | pg_attribute_printf(1, 2) pg_attribute_unused() static inline void VALGRIND_PRINTF(const char *format, ...) {}
      |                                                 ^~~~~~
make[1]: *** [<builtin>: pglogical_dependency.o] Error 1
In file included from pglogical_output_plugin.c:35:
pglogical.h:143:1: warning: ‘maybe_unused’ attribute ignored [-Wattributes]
  143 | pg_attribute_printf(1, 2) pg_attribute_unused() static inline void VALGRIND_PRINTF(const char *format, ...) {}
      | ^~~~~~~~~~~~~~~~~~~
pglogical.h:143:49: error: expected identifier or ‘(’ before ‘static’
  143 | pg_attribute_printf(1, 2) pg_attribute_unused() static inline void VALGRIND_PRINTF(const char *format, ...) {}
      |                                                 ^~~~~~
pglogical_output_plugin.c: In function ‘pg_decode_startup’:
pglogical_output_plugin.c:377:9: error: implicit declaration of function ‘VALGRIND_PRINTF’ [-Wimplicit-function-declaration]
  377 |         VALGRIND_PRINTF("PGLOGICAL: pglogical worker output_plugin\n");
      |         ^~~~~~~~~~~~~~~
make[1]: *** [<builtin>: pglogical_output_plugin.o] Error 1
In file included from pglogical_apply_spi.c:44:
pglogical.h:143:1: warning: ‘maybe_unused’ attribute ignored [-Wattributes]
  143 | pg_attribute_printf(1, 2) pg_attribute_unused() static inline void VALGRIND_PRINTF(const char *format, ...) {}
      | ^~~~~~~~~~~~~~~~~~~
pglogical.h:143:49: error: expected identifier or ‘(’ before ‘static’
  143 | pg_attribute_printf(1, 2) pg_attribute_unused() static inline void VALGRIND_PRINTF(const char *format, ...) {}
      |                                                 ^~~~~~
make[1]: *** [<builtin>: pglogical_apply_spi.o] Error 1
In file included from pglogical_worker.h:18,
                 from pglogical_sync.c:67:
pglogical.h:143:1: warning: ‘maybe_unused’ attribute ignored [-Wattributes]
  143 | pg_attribute_printf(1, 2) pg_attribute_unused() static inline void VALGRIND_PRINTF(const char *format, ...) {}
      | ^~~~~~~~~~~~~~~~~~~
pglogical.h:143:49: error: expected identifier or ‘(’ before ‘static’
  143 | pg_attribute_printf(1, 2) pg_attribute_unused() static inline void VALGRIND_PRINTF(const char *format, ...) {}
      |                                                 ^~~~~~
In file included from /usr/pgsql-19/include/server/postgres.h:49,
                 from pglogical_sync.c:14:
pglogical_sync.c: In function ‘pglogical_sync_subscription’:
/usr/pgsql-19/include/server/utils/elog.h:390:29: warning: declaration of ‘_save_exception_stack’ shadows a previous local [-Wshadow=compatible-local]
  390 |                 sigjmp_buf *_save_exception_stack##__VA_ARGS__ = PG_exception_stack; \
      |                             ^~~~~~~~~~~~~~~~~~~~~
/usr/pgsql-19/include/server/storage/ipc.h:50:17: note: in expansion of macro ‘PG_TRY’
   50 |                 PG_TRY()
      |                 ^~~~~~
pglogical_sync.c:941:25: note: in expansion of macro ‘PG_ENSURE_ERROR_CLEANUP’
  941 |                         PG_ENSURE_ERROR_CLEANUP(pglogical_sync_tmpfile_cleanup_cb,
      |                         ^~~~~~~~~~~~~~~~~~~~~~~
/usr/pgsql-19/include/server/utils/elog.h:390:29: note: shadowed declaration is here
  390 |                 sigjmp_buf *_save_exception_stack##__VA_ARGS__ = PG_exception_stack; \
      |                             ^~~~~~~~~~~~~~~~~~~~~
/usr/pgsql-19/include/server/storage/ipc.h:50:17: note: in expansion of macro ‘PG_TRY’
   50 |                 PG_TRY()
      |                 ^~~~~~
pglogical_sync.c:932:17: note: in expansion of macro ‘PG_ENSURE_ERROR_CLEANUP’
  932 |                 PG_ENSURE_ERROR_CLEANUP(pglogical_sync_worker_cleanup_error_cb,
      |                 ^~~~~~~~~~~~~~~~~~~~~~~
/usr/pgsql-19/include/server/utils/elog.h:391:39: warning: declaration of ‘_save_context_stack’ shadows a previous local [-Wshadow=compatible-local]
  391 |                 ErrorContextCallback *_save_context_stack##__VA_ARGS__ = error_context_stack; \
      |                                       ^~~~~~~~~~~~~~~~~~~
/usr/pgsql-19/include/server/storage/ipc.h:50:17: note: in expansion of macro ‘PG_TRY’
   50 |                 PG_TRY()
      |                 ^~~~~~
pglogical_sync.c:941:25: note: in expansion of macro ‘PG_ENSURE_ERROR_CLEANUP’
  941 |                         PG_ENSURE_ERROR_CLEANUP(pglogical_sync_tmpfile_cleanup_cb,
      |                         ^~~~~~~~~~~~~~~~~~~~~~~
/usr/pgsql-19/include/server/utils/elog.h:391:39: note: shadowed declaration is here
  391 |                 ErrorContextCallback *_save_context_stack##__VA_ARGS__ = error_context_stack; \
      |                                       ^~~~~~~~~~~~~~~~~~~
/usr/pgsql-19/include/server/storage/ipc.h:50:17: note: in expansion of macro ‘PG_TRY’
   50 |                 PG_TRY()
      |                 ^~~~~~
pglogical_sync.c:932:17: note: in expansion of macro ‘PG_ENSURE_ERROR_CLEANUP’
  932 |                 PG_ENSURE_ERROR_CLEANUP(pglogical_sync_worker_cleanup_error_cb,
      |                 ^~~~~~~~~~~~~~~~~~~~~~~
/usr/pgsql-19/include/server/utils/elog.h:392:28: warning: declaration of ‘_local_sigjmp_buf’ shadows a previous local [-Wshadow=compatible-local]
  392 |                 sigjmp_buf _local_sigjmp_buf##__VA_ARGS__; \
      |                            ^~~~~~~~~~~~~~~~~
/usr/pgsql-19/include/server/storage/ipc.h:50:17: note: in expansion of macro ‘PG_TRY’
   50 |                 PG_TRY()
      |                 ^~~~~~
pglogical_sync.c:941:25: note: in expansion of macro ‘PG_ENSURE_ERROR_CLEANUP’
  941 |                         PG_ENSURE_ERROR_CLEANUP(pglogical_sync_tmpfile_cleanup_cb,
      |                         ^~~~~~~~~~~~~~~~~~~~~~~
/usr/pgsql-19/include/server/utils/elog.h:392:28: note: shadowed declaration is here
  392 |                 sigjmp_buf _local_sigjmp_buf##__VA_ARGS__; \
      |                            ^~~~~~~~~~~~~~~~~
/usr/pgsql-19/include/server/storage/ipc.h:50:17: note: in expansion of macro ‘PG_TRY’
   50 |                 PG_TRY()
      |                 ^~~~~~
pglogical_sync.c:932:17: note: in expansion of macro ‘PG_ENSURE_ERROR_CLEANUP’
  932 |                 PG_ENSURE_ERROR_CLEANUP(pglogical_sync_worker_cleanup_error_cb,
      |                 ^~~~~~~~~~~~~~~~~~~~~~~
/usr/pgsql-19/include/server/utils/elog.h:393:22: warning: declaration of ‘_do_rethrow’ shadows a previous local [-Wshadow=compatible-local]
  393 |                 bool _do_rethrow##__VA_ARGS__ = false; \
      |                      ^~~~~~~~~~~
/usr/pgsql-19/include/server/storage/ipc.h:50:17: note: in expansion of macro ‘PG_TRY’
   50 |                 PG_TRY()
      |                 ^~~~~~
pglogical_sync.c:941:25: note: in expansion of macro ‘PG_ENSURE_ERROR_CLEANUP’
  941 |                         PG_ENSURE_ERROR_CLEANUP(pglogical_sync_tmpfile_cleanup_cb,
      |                         ^~~~~~~~~~~~~~~~~~~~~~~
/usr/pgsql-19/include/server/utils/elog.h:393:22: note: shadowed declaration is here
  393 |                 bool _do_rethrow##__VA_ARGS__ = false; \
      |                      ^~~~~~~~~~~
/usr/pgsql-19/include/server/storage/ipc.h:50:17: note: in expansion of macro ‘PG_TRY’
   50 |                 PG_TRY()
      |                 ^~~~~~
pglogical_sync.c:932:17: note: in expansion of macro ‘PG_ENSURE_ERROR_CLEANUP’
  932 |                 PG_ENSURE_ERROR_CLEANUP(pglogical_sync_worker_cleanup_error_cb,
      |                 ^~~~~~~~~~~~~~~~~~~~~~~
In file included from pglogical_worker.h:18,
                 from pglogical_apply.c:70:
pglogical.h:143:1: warning: ‘maybe_unused’ attribute ignored [-Wattributes]
  143 | pg_attribute_printf(1, 2) pg_attribute_unused() static inline void VALGRIND_PRINTF(const char *format, ...) {}
      | ^~~~~~~~~~~~~~~~~~~
pglogical.h:143:49: error: expected identifier or ‘(’ before ‘static’
  143 | pg_attribute_printf(1, 2) pg_attribute_unused() static inline void VALGRIND_PRINTF(const char *format, ...) {}
      |                                                 ^~~~~~
pglogical_apply.c: In function ‘handle_begin’:
pglogical_apply.c:299:9: error: implicit declaration of function ‘VALGRIND_PRINTF’ [-Wimplicit-function-declaration]
  299 |         VALGRIND_PRINTF("PGLOGICAL_APPLY: begin %u\n", remote_xid);
      |         ^~~~~~~~~~~~~~~
make[1]: *** [<builtin>: pglogical_sync.o] Error 1
make[1]: *** [<builtin>: pglogical_apply.o] Error 1
In file included from pglogical_worker.h:18,
                 from pglogical_worker.c:36:
pglogical.h:143:1: warning: ‘maybe_unused’ attribute ignored [-Wattributes]
  143 | pg_attribute_printf(1, 2) pg_attribute_unused() static inline void VALGRIND_PRINTF(const char *format, ...) {}
      | ^~~~~~~~~~~~~~~~~~~
pglogical.h:143:49: error: expected identifier or ‘(’ before ‘static’
  143 | pg_attribute_printf(1, 2) pg_attribute_unused() static inline void VALGRIND_PRINTF(const char *format, ...) {}
      |                                                 ^~~~~~
pglogical_worker.c: In function ‘pglogical_worker_attach’:
pglogical_worker.c:349:9: error: implicit declaration of function ‘VALGRIND_PRINTF’ [-Wimplicit-function-declaration]
  349 |         VALGRIND_PRINTF("PGLOGICAL: pglogical worker %s (%s)\n",
      |         ^~~~~~~~~~~~~~~
make[1]: *** [<builtin>: pglogical_worker.o] Error 1

```


That said these are the warnings against PostgreSQL 16 to 18 on the same platforms:

```
In file included from /usr/pgsql-19/include/server/postgres.h:49,
                 from pglogical_sync.c:14:
pglogical_sync.c: In function ‘pglogical_sync_subscription’:
/usr/pgsql-19/include/server/utils/elog.h:390:29: warning: declaration of ‘_save_exception_stack’ shadows a previous local [-Wshadow=compatible-local]
  390 |                 sigjmp_buf *_save_exception_stack##__VA_ARGS__ = PG_exception_stack; \
      |                             ^~~~~~~~~~~~~~~~~~~~~
/usr/pgsql-19/include/server/storage/ipc.h:50:17: note: in expansion of macro ‘PG_TRY’
   50 |                 PG_TRY()
      |                 ^~~~~~
pglogical_sync.c:941:25: note: in expansion of macro ‘PG_ENSURE_ERROR_CLEANUP’
  941 |                         PG_ENSURE_ERROR_CLEANUP(pglogical_sync_tmpfile_cleanup_cb,
      |                         ^~~~~~~~~~~~~~~~~~~~~~~
/usr/pgsql-19/include/server/utils/elog.h:390:29: note: shadowed declaration is here
  390 |                 sigjmp_buf *_save_exception_stack##__VA_ARGS__ = PG_exception_stack; \
      |                             ^~~~~~~~~~~~~~~~~~~~~
/usr/pgsql-19/include/server/storage/ipc.h:50:17: note: in expansion of macro ‘PG_TRY’
   50 |                 PG_TRY()
      |                 ^~~~~~
pglogical_sync.c:932:17: note: in expansion of macro ‘PG_ENSURE_ERROR_CLEANUP’
  932 |                 PG_ENSURE_ERROR_CLEANUP(pglogical_sync_worker_cleanup_error_cb,
      |                 ^~~~~~~~~~~~~~~~~~~~~~~
/usr/pgsql-19/include/server/utils/elog.h:391:39: warning: declaration of ‘_save_context_stack’ shadows a previous local [-Wshadow=compatible-local]
  391 |                 ErrorContextCallback *_save_context_stack##__VA_ARGS__ = error_context_stack; \
      |                                       ^~~~~~~~~~~~~~~~~~~
/usr/pgsql-19/include/server/storage/ipc.h:50:17: note: in expansion of macro ‘PG_TRY’
   50 |                 PG_TRY()
      |                 ^~~~~~
pglogical_sync.c:941:25: note: in expansion of macro ‘PG_ENSURE_ERROR_CLEANUP’
  941 |                         PG_ENSURE_ERROR_CLEANUP(pglogical_sync_tmpfile_cleanup_cb,
      |                         ^~~~~~~~~~~~~~~~~~~~~~~
/usr/pgsql-19/include/server/utils/elog.h:391:39: note: shadowed declaration is here
  391 |                 ErrorContextCallback *_save_context_stack##__VA_ARGS__ = error_context_stack; \
      |                                       ^~~~~~~~~~~~~~~~~~~
/usr/pgsql-19/include/server/storage/ipc.h:50:17: note: in expansion of macro ‘PG_TRY’
   50 |                 PG_TRY()
      |                 ^~~~~~
pglogical_sync.c:932:17: note: in expansion of macro ‘PG_ENSURE_ERROR_CLEANUP’
  932 |                 PG_ENSURE_ERROR_CLEANUP(pglogical_sync_worker_cleanup_error_cb,
      |                 ^~~~~~~~~~~~~~~~~~~~~~~
/usr/pgsql-19/include/server/utils/elog.h:392:28: warning: declaration of ‘_local_sigjmp_buf’ shadows a previous local [-Wshadow=compatible-local]
  392 |                 sigjmp_buf _local_sigjmp_buf##__VA_ARGS__; \
      |                            ^~~~~~~~~~~~~~~~~
/usr/pgsql-19/include/server/storage/ipc.h:50:17: note: in expansion of macro ‘PG_TRY’
   50 |                 PG_TRY()
      |                 ^~~~~~
pglogical_sync.c:941:25: note: in expansion of macro ‘PG_ENSURE_ERROR_CLEANUP’
  941 |                         PG_ENSURE_ERROR_CLEANUP(pglogical_sync_tmpfile_cleanup_cb,
      |                         ^~~~~~~~~~~~~~~~~~~~~~~
/usr/pgsql-19/include/server/utils/elog.h:392:28: note: shadowed declaration is here
  392 |                 sigjmp_buf _local_sigjmp_buf##__VA_ARGS__; \
      |                            ^~~~~~~~~~~~~~~~~
/usr/pgsql-19/include/server/storage/ipc.h:50:17: note: in expansion of macro ‘PG_TRY’
   50 |                 PG_TRY()
      |                 ^~~~~~
pglogical_sync.c:932:17: note: in expansion of macro ‘PG_ENSURE_ERROR_CLEANUP’
  932 |                 PG_ENSURE_ERROR_CLEANUP(pglogical_sync_worker_cleanup_error_cb,
      |                 ^~~~~~~~~~~~~~~~~~~~~~~
/usr/pgsql-19/include/server/utils/elog.h:393:22: warning: declaration of ‘_do_rethrow’ shadows a previous local [-Wshadow=compatible-local]
  393 |                 bool _do_rethrow##__VA_ARGS__ = false; \
      |                      ^~~~~~~~~~~
/usr/pgsql-19/include/server/storage/ipc.h:50:17: note: in expansion of macro ‘PG_TRY’
   50 |                 PG_TRY()
      |                 ^~~~~~
pglogical_sync.c:941:25: note: in expansion of macro ‘PG_ENSURE_ERROR_CLEANUP’
  941 |                         PG_ENSURE_ERROR_CLEANUP(pglogical_sync_tmpfile_cleanup_cb,
      |                         ^~~~~~~~~~~~~~~~~~~~~~~
/usr/pgsql-19/include/server/utils/elog.h:393:22: note: shadowed declaration is here
  393 |                 bool _do_rethrow##__VA_ARGS__ = false; \
      |                      ^~~~~~~~~~~
/usr/pgsql-19/include/server/storage/ipc.h:50:17: note: in expansion of macro ‘PG_TRY’
   50 |                 PG_TRY()
      |                 ^~~~~~
pglogical_sync.c:932:17: note: in expansion of macro ‘PG_ENSURE_ERROR_CLEANUP’
  932 |                 PG_ENSURE_ERROR_CLEANUP(pglogical_sync_worker_cleanup_error_cb,
      |                 ^~~~~~~~~~~~~~~~~~~~~~~

```
Can you please take a look? Thanks!

Regards, Devrim

Hi @eulerto ,

Thanks!  That PR fixes the build issue. Just getting the warnings against PG16:

```
gcc -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Werror=vla -Wmissing-format-attribute -Wold-style-declaration -Wimplicit-fallthrough=5 -Wcast-function-type -Wshadow=compatible-local -Wformat-security -Wmissing-variable-declarations -fno-strict-aliasing -fwrapv -fexcess-precision=standard -Wno-format-truncation -Wno-stringop-truncation -g -O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=3 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -march=x86-64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -mtls-dialect=gnu2 -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer -Wstrict-prototypes -Wold-style-definition -fPIC -fvisibility=hidden -fPIC -fvisibility=hidden -I/usr/pgsql-19/include -I/tmp/repo/pglogical/compat19 -Werror=implicit-function-declaration -I. -I./ -I/usr/pgsql-19/include/server -I/usr/pgsql-19/include/internal -D_GNU_SOURCE  -I/usr/include -I/usr/include/libxml2 -DWITH_GZFILEOP     -c -o pglogical_output.o pglogical_output.c
/usr/bin/clang -Wno-ignored-attributes -fno-strict-aliasing -fwrapv -fexcess-precision=standard -Wno-unused-command-line-argument -Wno-compound-token-split-by-macro -Wno-format-truncation -O2  -I/usr/pgsql-19/include -I/tmp/repo/pglogical/compat19 -Werror=implicit-function-declaration -I. -I./ -I/usr/pgsql-19/include/server -I/usr/pgsql-19/include/internal -D_GNU_SOURCE  -I/usr/include -I/usr/include/libxml2 -DWITH_GZFILEOP    -flto=thin -emit-llvm -c -o pglogical_output.bc pglogical_output.c
/usr/bin/clang -Wno-ignored-attributes -fno-strict-aliasing -fwrapv -fexcess-precision=standard -Wno-unused-command-line-argument -Wno-compound-token-split-by-macro -Wno-format-truncation -O2  -I/usr/pgsql-19/include -I/tmp/repo/pglogical/compat19 -Werror=implicit-function-declaration -I. -I./ -I/usr/pgsql-19/include/server -I/usr/pgsql-19/include/internal -D_GNU_SOURCE  -I/usr/include -I/usr/include/libxml2 -DWITH_GZFILEOP    -flto=thin -emit-llvm -c -o pglogical_apply.bc pglogical_apply.c
/usr/bin/clang -Wno-ignored-attributes -fno-strict-aliasing -fwrapv -fexcess-precision=standard -Wno-unused-command-line-argument -Wno-compound-token-split-by-macro -Wno-format-truncation -O2  -I/usr/pgsql-19/include -I/tmp/repo/pglogical/compat19 -Werror=implicit-function-declaration -I. -I./ -I/usr/pgsql-19/include/server -I/usr/pgsql-19/include/internal -D_GNU_SOURCE  -I/usr/include -I/usr/include/libxml2 -DWITH_GZFILEOP    -flto=thin -emit-llvm -c -o pglogical_conflict.bc pglogical_conflict.c
In file included from /usr/pgsql-19/include/server/postgres.h:49,
                 from pglogical_sync.c:14:
pglogical_sync.c: In function ‘pglogical_sync_subscription’:
/usr/pgsql-19/include/server/utils/elog.h:390:29: warning: declaration of ‘_save_exception_stack’ shadows a previous local [-Wshadow=compatible-local]
  390 |                 sigjmp_buf *_save_exception_stack##__VA_ARGS__ = PG_exception_stack; \
      |                             ^~~~~~~~~~~~~~~~~~~~~
/usr/pgsql-19/include/server/storage/ipc.h:50:17: note: in expansion of macro ‘PG_TRY’
   50 |                 PG_TRY()
      |                 ^~~~~~
pglogical_sync.c:941:25: note: in expansion of macro ‘PG_ENSURE_ERROR_CLEANUP’
  941 |                         PG_ENSURE_ERROR_CLEANUP(pglogical_sync_tmpfile_cleanup_cb,
      |                         ^~~~~~~~~~~~~~~~~~~~~~~
/usr/pgsql-19/include/server/utils/elog.h:390:29: note: shadowed declaration is here
  390 |                 sigjmp_buf *_save_exception_stack##__VA_ARGS__ = PG_exception_stack; \
      |                             ^~~~~~~~~~~~~~~~~~~~~
/usr/pgsql-19/include/server/storage/ipc.h:50:17: note: in expansion of macro ‘PG_TRY’
   50 |                 PG_TRY()
      |                 ^~~~~~
pglogical_sync.c:932:17: note: in expansion of macro ‘PG_ENSURE_ERROR_CLEANUP’
  932 |                 PG_ENSURE_ERROR_CLEANUP(pglogical_sync_worker_cleanup_error_cb,
      |                 ^~~~~~~~~~~~~~~~~~~~~~~
/usr/pgsql-19/include/server/utils/elog.h:391:39: warning: declaration of ‘_save_context_stack’ shadows a previous local [-Wshadow=compatible-local]
  391 |                 ErrorContextCallback *_save_context_stack##__VA_ARGS__ = error_context_stack; \
      |                                       ^~~~~~~~~~~~~~~~~~~
/usr/pgsql-19/include/server/storage/ipc.h:50:17: note: in expansion of macro ‘PG_TRY’
   50 |                 PG_TRY()
      |                 ^~~~~~
pglogical_sync.c:941:25: note: in expansion of macro ‘PG_ENSURE_ERROR_CLEANUP’
  941 |                         PG_ENSURE_ERROR_CLEANUP(pglogical_sync_tmpfile_cleanup_cb,
      |                         ^~~~~~~~~~~~~~~~~~~~~~~
/usr/pgsql-19/include/server/utils/elog.h:391:39: note: shadowed declaration is here
  391 |                 ErrorContextCallback *_save_context_stack##__VA_ARGS__ = error_context_stack; \
      |                                       ^~~~~~~~~~~~~~~~~~~
/usr/pgsql-19/include/server/storage/ipc.h:50:17: note: in expansion of macro ‘PG_TRY’
   50 |                 PG_TRY()
      |                 ^~~~~~
pglogical_sync.c:932:17: note: in expansion of macro ‘PG_ENSURE_ERROR_CLEANUP’
  932 |                 PG_ENSURE_ERROR_CLEANUP(pglogical_sync_worker_cleanup_error_cb,
      |                 ^~~~~~~~~~~~~~~~~~~~~~~
/usr/pgsql-19/include/server/utils/elog.h:392:28: warning: declaration of ‘_local_sigjmp_buf’ shadows a previous local [-Wshadow=compatible-local]
  392 |                 sigjmp_buf _local_sigjmp_buf##__VA_ARGS__; \
      |                            ^~~~~~~~~~~~~~~~~
/usr/pgsql-19/include/server/storage/ipc.h:50:17: note: in expansion of macro ‘PG_TRY’
   50 |                 PG_TRY()
      |                 ^~~~~~
pglogical_sync.c:941:25: note: in expansion of macro ‘PG_ENSURE_ERROR_CLEANUP’
  941 |                         PG_ENSURE_ERROR_CLEANUP(pglogical_sync_tmpfile_cleanup_cb,
      |                         ^~~~~~~~~~~~~~~~~~~~~~~
/usr/pgsql-19/include/server/utils/elog.h:392:28: note: shadowed declaration is here
  392 |                 sigjmp_buf _local_sigjmp_buf##__VA_ARGS__; \
      |                            ^~~~~~~~~~~~~~~~~
/usr/pgsql-19/include/server/storage/ipc.h:50:17: note: in expansion of macro ‘PG_TRY’
   50 |                 PG_TRY()
      |                 ^~~~~~
pglogical_sync.c:932:17: note: in expansion of macro ‘PG_ENSURE_ERROR_CLEANUP’
  932 |                 PG_ENSURE_ERROR_CLEANUP(pglogical_sync_worker_cleanup_error_cb,
      |                 ^~~~~~~~~~~~~~~~~~~~~~~
/usr/pgsql-19/include/server/utils/elog.h:393:22: warning: declaration of ‘_do_rethrow’ shadows a previous local [-Wshadow=compatible-local]
  393 |                 bool _do_rethrow##__VA_ARGS__ = false; \
      |                      ^~~~~~~~~~~
/usr/pgsql-19/include/server/storage/ipc.h:50:17: note: in expansion of macro ‘PG_TRY’
   50 |                 PG_TRY()
      |                 ^~~~~~
pglogical_sync.c:941:25: note: in expansion of macro ‘PG_ENSURE_ERROR_CLEANUP’
  941 |                         PG_ENSURE_ERROR_CLEANUP(pglogical_sync_tmpfile_cleanup_cb,
      |                         ^~~~~~~~~~~~~~~~~~~~~~~
/usr/pgsql-19/include/server/utils/elog.h:393:22: note: shadowed declaration is here
  393 |                 bool _do_rethrow##__VA_ARGS__ = false; \
      |                      ^~~~~~~~~~~
/usr/pgsql-19/include/server/storage/ipc.h:50:17: note: in expansion of macro ‘PG_TRY’
   50 |                 PG_TRY()
      |                 ^~~~~~
pglogical_sync.c:932:17: note: in expansion of macro ‘PG_ENSURE_ERROR_CLEANUP’
  932 |                 PG_ENSURE_ERROR_CLEANUP(pglogical_sync_worker_cleanup_error_cb,
      |                 ^~~~~~~~~~~~~~~~~~~~~~~

```

--- a/pglogical.h
+++ b/pglogical.h
@@ -140,7 +140,7 @@ extern bool in_pglogical_replicate_ddl_command;
  * the worker type, etc, prefixed by PGLOGICAL:, in the Valgrind logs. We
  * need to stub it out if we aren't using valgrind.
  */
-static inline void pg_attribute_printf(1, 2) pg_attribute_unused() VALGRIND_PRINTF(const char *format, ...) {}
+static inline void pg_attribute_printf(1, 2) VALGRIND_PRINTF(const char *format, ...) {}
 
 #endif
 

--- a/pglogical.h
+++ b/pglogical.h
@@ -140,7 +140,7 @@ extern bool in_pglogical_replicate_ddl_command;
  * the worker type, etc, prefixed by PGLOGICAL:, in the Valgrind logs. We
  * need to stub it out if we aren't using valgrind.
  */
-pg_attribute_printf(1, 2) pg_attribute_unused() static inline void VALGRIND_PRINTF(const char *format, ...) {}
+static inline void pg_attribute_printf(1, 2) pg_attribute_unused() VALGRIND_PRINTF(const char *format, ...) {}
 
 #endif
 

